internal/service/resource_fetcher.go

```diff
@@ -0,0 +1,302 @@
+	value string) *ResourceFetcherBuilder {
+	b.cloudID = value
+	return b
+}
+
+// SetBackendURL sets the URL of the backend server This is mandatory.
+func (b *ResourceFetcherBuilder) SetBackendToken(
+	value string) *ResourceFetcherBuilder {
+	b.backendToken = value
+	return b
+}
+
+// SetBackendToken sets the authentication token that will be used to authenticate to the backend
+// server. This is mandatory.
+func (b *ResourceFetcherBuilder) SetBackendURL(
+	value string) *ResourceFetcherBuilder {
+	b.backendURL = value
+	return b
+}
+
+// SetGraphqlQuery sets the query to send to the search API server. This is mandatory.
+func (b *ResourceFetcherBuilder) SetGraphqlQuery(
+	value string) *ResourceFetcherBuilder {
+	b.graphqlQuery = value
+	return b
+}
+
+// SetGraphqlVars sets the query vars to send to the search API server. This is mandatory.
+func (b *ResourceFetcherBuilder) SetGraphqlVars(
+	value *model.SearchInput) *ResourceFetcherBuilder {
+	b.graphqlVars = value
+	return b
+}
+
+// Build uses the data stored in the builder to create and configure a new handler.
+func (b *ResourceFetcherBuilder) Build() (
+	result *ResourceFetcher, err error) {
+	// Check parameters:
+	if b.logger == nil {
+		err = errors.New("logger is mandatory")
+		return
+	}
+	if b.cloudID == "" {
+		err = errors.New("cloud identifier is mandatory")
+		return
+	}
+	if b.backendURL == "" {
+		err = errors.New("backend URL is mandatory")
+		return
+	}
+	if b.backendToken == "" {
+		err = errors.New("backend token is mandatory")
+		return
+	}
+
+	// Create the HTTP client that we will use to connect to the backend:
+	var backendTransport http.RoundTripper
+	backendTransport = &http.Transport{
+		TLSClientConfig: &tls.Config{
+			InsecureSkipVerify: true,
+		},
+	}
+	if b.transportWrapper != nil {
+		backendTransport = b.transportWrapper(backendTransport)
+	}
+	backendClient := &http.Client{
+		Transport: backendTransport,
+	}
+
+	// Create and populate the object:
+	result = &ResourceFetcher{
+		logger:        b.logger,
+		cloudID:       b.cloudID,
+		backendURL:    b.backendURL,
+		backendToken:  b.backendToken,
+		backendClient: backendClient,
+		graphqlQuery:  b.graphqlQuery,
+		graphqlVars:   b.graphqlVars,
+	}
+	return
+}
+
+// FetchItems returns a data stream of O2 Resources.
+// The items are converted from Nodes fetched from the search API.
+func (r *ResourceFetcher) FetchItems(
+	ctx context.Context, resourcePoolID string) (resources data.Stream, err error) {
+	// Search Nodes
+	resultArr, err := r.getSearchResults(ctx)
+	if err != nil {
+		return
+	}
+
+	// Extract 'searchResult' object
+	searchResult := resultArr[0].(map[string]any)
+
+	// Convert response to json
+	items, err := json.Marshal(searchResult)
+	if err != nil {
+		return
+	}
+	itemsReader := bytes.NewReader(items)
+
+	// Create reader for Nodes
+	nodes, err := k8s.NewStream().
+		SetLogger(r.logger).
+		SetReader(itemsReader).
+		Build()
+	if err != nil {
+		return
+	}
```

Comment:
I think this conversion to JSON in order to use the "k8s.NewStream" isn't necessary. If you already have an slice containing the input objects then you can create a stream from it using the [data.Pour](https://github.com/openshift-kni/oran-o2ims/blob/45562d9bc2afa0ff70b6d353dd0aaa53e5f0fc8f/internal/data/data.go#L90-L91) function:

```go
resources = data.Map(data.Pour(resultArr), r.mapNodeItems)
```

However having that `resultArr` slice is an indication that we aren't taking advantage of the streaming approach. If you look at the `k8s.Stream` type you will see that what it reads the input until it finds the "items" element, and then it reads those items one by one, without storing them in memory. Would it be possible to do the same for the "searchResults" field returned by the GraphQL server? That would reduce the number of JSON conversions that we have to do, and would also save memory, as only one item would need to be in memory at a time.

If you agree with that approach we can do it in a separate pull request. I'd rather merge this as is.

Reply:
sure, will do.